### ISSUE TITLE
 Token type not implemented for brew pipe

### ISSUE THREAD

**Issue by deepak0405** [goreleaser/goreleaser#1119]
**Describe the bug**
This is not a bug, I am facing some error and could not find it anywhere in github issues or documentation. Please guide me how to solve this error, what is the cause of the error. Link to documentation will also be helpful.

**To Reproduce**
error=homebrew tap formula: failed to publish artifacts: token type not implemented for brew pipe
For the github token I only selected `repo` scope

```yaml
# This is an example goreleaser.yaml file with some sane defaults.
# Make sure to check the documentation at http://goreleaser.com
before:
  hooks:
    # you may remove this if you don't need go generate
    - go generate ./...
builds:
- env:
  - CGO_ENABLED=0
archives:
- replacements:
    darwin: Darwin
    linux: Linux
    windows: Windows
    386: i386
    amd64: x86_64
checksum:
  name_template: 'checksums.txt'
snapshot:
  name_template: "{{ .Tag }}-next"
changelog:
  sort: asc
  filters:
    exclude:
    - '^docs:'
    - '^test:'
# .goreleaser.yml
release:
  # Repo in which the release will be created.
  # Default is extracted from the origin remote URL.
  # Note: it can only be one: either github or gitlab
  github:
      owner: CompanyName
      name: myRepo

  # If set to true, will not auto-publish the release.
  # Default is false.
  draft: false

  # If set to auto, will mark the release as not ready for production
  # in case there is an indicator for this in the tag e.g. v1.0.0-rc1
  # If set to true, will mark the release as not ready for production.
  # Default is false.
  prerelease: auto

  # You can change the name of the GitHub release.
  # Default is `{{.Tag}}`
  name_template: "{{.ProjectName}}-v{{.Version}} {{.Env.USER}}"

  # You can disable this pipe in order to not upload any artifacts to
  # GitHub.
  # Defaults to false.
  disable: true
brews:
  -
    # Name template of the recipe
    # Default to project name
    name: someName

    # IDs of the archives to use.
    # Defaults to all.
    ids:


    # NOTE: make sure the url_template, the token and given repo (github or gitlab) owner and name are from the
    # same kind. We will probably unify this in the next major version like it is done with scoop.

    # Github repository to push the tap to.
    github:
      owner: CompanyName
      name: myRepo

    # Template for the url which is determined by the given Token (github or gitlab)
    # Default for github is "https://github.com/<repo_owner>/<repo_name>/releases/download/{{ .Tag }}/{{ .ArtifactName }}"
    # Default for gitlab is "https://gitlab.com/<repo_owner>/<repo_name>/uploads/{{ .ArtifactUploadHash }}/{{ .ArtifactName }}"
    url_template: "http://github.mycompany.com/foo/bar/releases/{{ .Tag }}/{{ .ArtifactName }}"

    # Allows you to set a custom download strategy.
    # Default is empty.
    download_strategy: GitHubPrivateRepositoryReleaseDownloadStrategy

    # Allows you to add a custom require_relative at the top of the formula template
    # Default is empty
    custom_require: 

    # Git author used to commit to the repository.
    # Defaults are shown.
    commit_author:
      name: deepak
      email: <email-id>

    # Folder inside the repository to put the formula.
    # Default is the root folder.
    folder: formula

    # Caveats for the user of your binary.
    # Default is empty.
    caveats: "Company Cmd Utility"

    # Your app's homepage.
    # Default is empty.
    homepage: "https://<company-homepage>/"

    # Your app's description.
    # Default is empty.
    description: "CompanyName"

    # Setting this will prevent goreleaser to actually try to commit the updated
    # formula - instead, the formula file will be stored on the dist folder only,
    # leaving the responsibility of publishing it to the user.
    # If set to auto, the release will not be uploaded to the homebrew tap
    # in case there is an indicator for prerelease in the tag e.g. v1.0.0-rc1
    # Default is false.
    skip_upload: false

    # Custom block for brew.
    # Can be used to specify alternate downloads for devel or head releases.
    # Default is empty.
    custom_block:

    # Packages your package depends on.
    dependencies:

    # Packages that conflict with your package.
    conflicts:

    # Specify for packages that run as a service.
    # Default is empty.
    plist:

    # So you can `brew test` your formula.
    # Default is empty.
    test: |
      system "#{bin}/companyName --version"
      ...

    # Custom install script for brew.
    # Default is 'bin.install "program"'.
    install: |
      bin.install "company"
      ...

```
Config is valid. I changed the name of my company at few places for creating issue.

**Environment (please complete the following information):**

- OS:  mac
- OS version: 10.14.6
- GoReleaser Version 0.116.0

**Additional context**

**Comment by mavogel** [comment]
Hi, could you provide more information? :)
- Does the release pipe pass without a problem?
- How do you set your `github` token? Env or file?

**Comment by deepak0405** [comment]
1. The release pipe fails
```
   • releasing using goreleaser 0.116.0...
   • loading config file       file=.goreleaser.yml
   • RUNNING BEFORE HOOKS
      • running go generate ./...
   • LOADING ENVIRONMENT VARIABLES
      • pipe skipped              error=release pipe is disabled
   • GETTING AND VALIDATING GIT STATE
      • releasing v1.0.0-rc7, commit befa21a42a5d2caf8c150145b1tbf9a7cda36f5d
   • PARSING TAG      
   • SETTING DEFAULTS 
      • LOADING ENVIRONMENT VARIABLES
      • SNAPSHOTING      
      • GITHUB/GITLAB RELEASES
      • PROJECT NAME     
      • BUILDING BINARIES
      • ARCHIVES         
      • LINUX PACKAGES WITH NFPM
      • SNAPCRAFT PACKAGES
      • CALCULATING CHECKSUMS
      • SIGNING ARTIFACTS
      • DOCKER IMAGES    
      • ARTIFACTORY      
      • S3               
      • BLOB             
      • HOMEBREW TAP FORMULA
      • SCOOP MANIFEST   
   • SNAPSHOTING      
      • pipe skipped              error=not a snapshot
   • CHECKING ./DIST  
      • --rm-dist is set, cleaning it up
   • WRITING EFFECTIVE CONFIG FILE
      • writing                   config=dist/config.yaml
   • GENERATING CHANGELOG
      • writing                   changelog=dist/CHANGELOG.md
   • BUILDING BINARIES
      • building                  binary=dist/companyNamecli_darwin_386/cli
      • building                  binary=dist/companyNamecli_darwin_amd64/cli
      • building                  binary=dist/companyNamecli_linux_386/cli
      • building                  binary=dist/companyNamecli_linux_amd64/cli
   • ARCHIVES         
      • creating                  archive=dist/companyNamecli_1.0.0-rc7_Linux_i386.tar.gz
      • creating                  archive=dist/companyNamecli_1.0.0-rc7_Darwin_i386.tar.gz
      • creating                  archive=dist/companyNamecli_1.0.0-rc7_Darwin_x86_64.tar.gz
      • creating                  archive=dist/companyNamecli_1.0.0-rc7_Linux_x86_64.tar.gz
   • LINUX PACKAGES WITH NFPM
      • pipe skipped              error=no output formats configured
   • SNAPCRAFT PACKAGES
      • pipe skipped              error=no summary nor description were provided
   • CALCULATING CHECKSUMS
      • checksumming              file=companyNamecli_1.0.0-rc7_Linux_x86_64.tar.gz
      • checksumming              file=companyNamecli_1.0.0-rc7_Linux_i386.tar.gz
      • checksumming              file=companyNamecli_1.0.0-rc7_Darwin_i386.tar.gz
      • checksumming              file=companyNamecli_1.0.0-rc7_Darwin_x86_64.tar.gz
   • SIGNING ARTIFACTS
      • pipe skipped              error=artifact signing is disabled
   • DOCKER IMAGES    
      • pipe skipped              error=docker section is not configured
   • PUBLISHING       
      • S3               
         • pipe skipped              error=s3 section is not configured
      • BLOB             
         • pipe skipped              error=Blob section is not configured
      • HTTP PUT         
         • pipe skipped              error=put section is not configured
      • ARTIFACTORY      
         • pipe skipped              error=artifactory section is not configured
      • DOCKER IMAGES    
      • SNAPCRAFT PACKAGES
      • GITHUB/GITLAB RELEASES
         • pipe skipped              error=release pipe is disabled
      • HOMEBREW TAP FORMULA
         • writing                   formula=dist/companyName.rb
   ⨯ release failed after 5.18s error=homebrew tap formula: failed to publish artifacts: token type not implemented for brew pipe
```
2. I set the environment using `export`

**Comment by deepak0405** [comment]
My .rb file was incorrect earlier.  The url of the release was wrong. Now I corrected it and I can download the achieve using the link. Still I am getting the same error. My repo is private and I used the following [tutorial](https://help.github.com/en/articles/creating-a-personal-access-token-for-the-command-line) to generate the token.
```
class CompanyName < Formula
  desc "CompanyName"
  homepage "https://<company's-url>/"
  version "1.0.0-rc10"
  bottle :unneeded

  if OS.mac?
    url "https://github.com/<org-name>/<repo-name>/archive/v1.0.0-rc10.tar.gz", :using => GitHubPrivateRepositoryReleaseDownloadStrategy
    sha256 "c53ecb1854bb0ba78f47b9233c94b140f70f77917985df2d16bead9a5208eee9"
  elsif OS.linux?
    if Hardware::CPU.intel?
      url "https://github.com//<org-name>/<repo-name>/archive/v1.0.0-rc10.tar.gz", :using => GitHubPrivateRepositoryReleaseDownloadStrategy
      sha256 "380b30a7d18972dd444379639d8f9ed1c5686ca8d581715b5f5057bb0c5ac5c4"
    end
  end

  def install
    bin.install "harness"
    ...
  end

  def caveats; <<~EOS
    Harness Cmd Utility
  EOS
  end

  test do
    system "#{bin}/harness --version"
    ...
  end
end
```

**Comment by mavogel** [comment]
Hey, ok I did not look closely enough at your configuration.

## Explanation
With your config (comments omitted)
```yaml
release:
  github:
      owner: CompanyName
      name: myRepo
  draft: false
  prerelease: auto
  name_template: "{{.ProjectName}}-v{{.Version}} {{.Env.USER}}"
  disable: true # <= you disable it here
```
which is also shown in the log
```
      • GITHUB/GITLAB RELEASES
         • pipe skipped              error=release pipe is disabled
```

As you see in the last line the `release` stage is disable but stages like `brew/scoop` depend on this stage because they refer to the uploaded artifacts in their **recipes** as shown below:

```
...
if OS.mac?
    url "https://github.com/<org-name>/<repo-name>/archive/v1.0.0-rc10.tar.gz", :using => GitHubPrivateRepositoryReleaseDownloadStrategy
    sha256 "c53ecb1854bb0ba78f47b9233c94b140f70f77917985df2d16bead9a5208eee9"
....
```

And I checked the code and it was already a known bug: https://github.com/goreleaser/goreleaser/blob/master/internal/pipe/brew/brew.go#L122

### Fix
- we need to update brew and scoop logic
- additionally, update the validate config function because it should not be possible to have pipe enabled which depend on other pipes

Sorry that this is not clear from the point of creating the config. We will provide a fix and clarification on that. Thanks for pointing this one out @deepak0405  :)

**Comment by deepak0405** [comment]
Hey @mavogel, Thanks for the quick and clear reply. The problem is solved.

**Comment by caarlos0** [comment]
I think we can leave this open :)

**Comment by caarlos0** [comment]
about `GitHubPrivateRepositoryReleaseDownloadStrategy`, it was removed from homebrew I'm afraid...


check https://github.com/goreleaser/goreleaser/issues/967

maybe we need to handle this better as well 🤔

**Comment by deepak0405** [comment]
Thanks @caarlos0 I removed `GitHubPrivateRepositoryReleaseDownloadStrategy` from my code.

**Comment by mavogel** [comment]
fixed by #1128.  The `GitHubPrivateRepositoryReleaseDownloadStrategy` will be addressed in #967

**Comment by M-050** [comment]
اﺑدمانساک♡...'M
